### ISSUE TITLE
Using same parameter names in trait functions does not cause an error until its used in impl block

### ISSUE THREAD

**Issue by psvri** [rust-lang/rust#108642]
### Code

```Rust
#![allow(unused)]
pub trait Temp {
    fn a(data: Self, data: Self);
}

/*impl Temp for i64 {
    fn a(data: Self, data: Self) {}
}*/
fn main() {

}
```

### Current output

```Shell
No output
```

### Desired output

```Shell
fn a(data: Self, data: Self) {}
                 ^^^^ used as parameter more than once
```

### Rationale and extra context

The above code complies successfully with no errors. But when I use the trait in the impl block I get the error that the parameter is used more than once. 

The reason this error is required is so that if libraries export such a trait then it wont be caught unless its used by the dependents.

### Other cases

```Rust
Playground link

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3c31b05ebd73f136de52f37107e62226
```

**Comment by euclio** [comment]
I'd like to work on this, but it seems like it would need to be a future-incompatible lint, since it compiles today.
